[MRG+1]: IO support for artemis123

bloyl opened the pull request:
This is a WIP to add IO support for the artemis123 MEG machine by Tristan Technologies.

Known issues.
  1) Reference channel Location
  2) Coil definitions (both sensor and reference need to be changed)

Possible issues
  1) Added new FIFF constants to represent the sensors. Numbers may/may not conflict with other constants chosen else where in mne-python.

Eric89GXL commented:
One of your `git` accounts is mis-configured -- check the commit list on GitHub only some are attributed to you.

Regarding plotting using `plot_trans`, do you have any idea what your coils actually look like? We could easily make them circles or squares, or more complex shapes.

The reference channel locations would be useful, but IIRC they don't exist for some other systems (KIT maybe?) so it's not the end of the world if you don't have those to start.

Can you record some tiny (a few seconds) of empty-room data from the system? Then we can add it to the mne-testing-data repository for testing purposes.

bloyl commented:
is the git account thing causing issue? I had to switch to a non-development box to push to github and the git configs are different. is this what is causing travis to not work?

all of my coils are circles so I think I added it to the correct case in plot_trans. What needs to be updated in coil_defs.dat is the coil diameter and integration points (at the moment they are still the CTF coils defs) This is something I can fix once I think the more important things are in place.

I added some data (3sec) for testing purposes to
mne/io/artemis123/tests/data/Artemis_Data_2016-11-03-15h-58m_test.[txt/bin]
which should be part of this PR.

read_raw_artemis123 wants the bin file.

agramfort commented:
I'd recommend to put the data in the testing dataset to avoid having a growing github repo.

bloyl commented:
That's not a problem. Are there instructions on how to do that somewhere. I don't see anything on the contribute page but maybe I missed it.

Eric89GXL commented:
I don't think we have any instructions for that. Just make a PR to mne-tools/mne-testing-data, maybe inside a new `artemis123` directory. Then you'll need to delete the file from this PR and squash the commit(s) so that it is removed from the history.

bloyl commented:
Testing data is in WIP: Artemis123 datasets for IO testing #17 of mne-tools/mne-testing-data.

Note that artemis123 tests will fail unless the mne-testing-data PR 17 is used.

bloyl commented:
@Eric89GXL I don't understand why travis-ci builds are failing? it doesn't seem like something in the PR? Can you also see if I have successfully squashed all references to the data files. I think I have.

Eric89GXL commented on mne/viz/_3d.py:
```diff
@@ -584,7 +584,7 @@ def _sensor_shape(coil):
         rrs = np.array([circle.real, circle.imag]).T
         tris = _make_tris_fan(n_pts + 1)
-    elif id_ in (4002, 5001, 5003, 5004, 4004, 4005, 6001, 7001):
+    elif id_ in (4002, 5001, 5003, 5004, 4004, 4005, 6001, 7001, 10001, 10002):
```
This would be more DRY and clearer to use the FIFF constant name, but since we were lazy about it before you don't have to change it

Eric89GXL commented on mne/io/artemis123/utils.py:
```diff
@@ -0,0 +1,90 @@
+    loc[3:13] = R.T.reshape(9)
+    return loc
+
+def _compute_rot(zAxis_dev):
```
We don't use camelCase for functions or variables, can you change it to use underscore names?

Eric89GXL replied:
If this isn't in our contribution docs, I apologize for the extra time... feel free to update the docs too (in this PR or separately) or I can do it

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,217 @@
+    #build mne info struct
+    info = _empty_info(float(header_info['Rate Out']))
+
+    #Attempt to get time/date from fname
```
Is this standard imposed by the acquisition software, or local convention? Please add a comment specifying

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,217 @@
+    #Attempt to get time/date from fname
+    try:
+        date = datetime.datetime.strptime(op.basename(fname).split('_')[2],'%Y-%m-%d-%Hh-%Mm')
+    except:
```
Do not `except:` but `except Exception:` because the former is too broad (will catch SystemExit)

Eric89GXL commented:
The `setup.py` changes would fix some of the Travis errors, I'm looking into the others

Eric89GXL commented:
@bloyl no need to change setup.py anymore, a rebase should fix the CIs

Eric89GXL commented:
@jaeilepp we need new FIFF constants for Artemis123 sensor types. Do you have time to clear it with the Elekta folks? We might want to rekindle the discussion of moving the constants to GitHub to facilitate things.

Eric89GXL commented:
It looks like commit https://github.com/mne-tools/mne-python/pull/3754/commits/49519a392bbef1184e7d2c2b48d8a24e45bbf52b has not been squashed out, but as long as we merge-by-squash (as we have been doing lately) we should be okay in terms of repo history.

It looks like the Travis builds are actually running tests now, which bodes well. Hopefully there are no failures, and if they are any, they are meaningful.

bloyl commented:
It might be worth a wider discussion on FIFF constants. I noticed there are at least 2 coil types 8001 and 9001 that are in coil_defs.dat and not in FIFF constants. I imagine there might be some other constants that are used but not defined in other places.

I also have to add an additional coil type to the 2 I have already added so artemis will have 10001,10002 and 10003. I'll try to push a commit this afternoon with all of the reference channels fully described.

Eric89GXL commented:
Lots of little style convention issues:

https://travis-ci.org/mne-tools/mne-python/jobs/175718178#L3121

And then you need to add your `.csv` file to `setup.py` in `package_data`:

https://travis-ci.org/mne-tools/mne-python/jobs/175718179#L2818

Eric89GXL commented:
BTW you can do `make flake`, `make docstyle` locally to test style issues

Eric89GXL commented:
FYI @bloyl it looks like Flake8 has uncovered a bunch of previously-skipped E305 errors, you can ignore those outside of your changes (I will fix them separately)

Eric89GXL commented:
More style stuff:

https://travis-ci.org/mne-tools/mne-python/jobs/175748120#L3133

Eric89GXL commented:
Can you add your system to the `plot_trans` system example? That way CircleCI will build it for us and we can see a nice illustration of the geometry.

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,258 @@
+def read_raw_artemis123(input_fname, preload=False, verbose=None):
+    """Read Artemis123 data as raw object.
+
+    Note: This reader takes data files with the extension ``.bin`` as an
```
`.. note::` see how we do it elsewhere. Although I don't think this even needs to have the `Note:` or `.. note::` prefix at all, so might as well remove it.

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,258 @@
+
+    Note: This reader takes data files with the extension ``.bin`` as an
+    input. The header file with the same file name stem and an extension
+    ``.`` is expected to be found in the same directory.
```
`.txt` right?

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,258 @@
+    # build description
+    desc = ''
+    for k in ['Project Name', 'Purpose', 'Notes', 'Subject ID']:
+        desc += '{} : {}\n'.format(k, header_info[k])
```
`info` already has a `info['subject_info']` so better to stash that in there

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,258 @@
+             'unit_mul': FIFF.FIFF_UNITM_NONE,
+             'coord_frame': FIFF.FIFFV_COORD_DEVICE}
+        try:
+            t['loc'] = loc_dict[chan['name']]
```
No need for try:
```
t['loc'] = loc_dict.get(chan['name'], np.zeros(12))
```

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,258 @@
+    # reduce info['bads'] to unique set
+    info['bads'] = list(set(info['bads']))
+    info['ch_names'] = [ch['ch_name'] for ch in info['chs']]
+    info['nchan'] = len(info['ch_names'])
```
Get rid of these last two, use `info._update_redundant()`. If this was copied from another reader, please update there, too...

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,258 @@
+    """
+
+    def __init__(self, input_fname, preload=False, verbose=None):  # noqa: D102
+        input_fname = op.abspath(input_fname)
```
might as well do this in `_get_artemis123_info`

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,258 @@
+        last_samps = [header_info['num_samples'] - 1]
+        super(RawArtemis123, self).__init__(
+            info, preload, filenames=[input_fname], raw_extras=[header_info],
+            last_samps=last_samps, orig_format='int',
```
`int`? it looks like `float`?

Eric89GXL commented on mne/io/artemis123/tests/test_artemis123.py:
```diff
@@ -0,0 +1,23 @@
+def test_data():
+    """Test reading raw Artemis123 files."""
+    _test_raw_reader(read_raw_artemis123, input_fname=short_no_HPI_fname)
+    pass
```
Don't need `pass` line here

Might be worth adding some sanity checks. If you have a MATLAB reader written already, for example, you could hard-code that some random time point for some random set of 10 channels gives an identical result (to numerical precision, i.e. `from numpy.testing import assert_allclose`)

Eric89GXL commented on mne/io/artemis123/utils.py:
```diff
@@ -0,0 +1,99 @@
+    """Load MNE locs stucture from file (if exists) or recreate it."""
+    # find input file
+    FILE = inspect.getfile(inspect.currentframe())
+    resource_dir = op.join(op.dirname(op.abspath(FILE)), 'resources')
```
you should be able to just do `op.abspath(__file__)`

Eric89GXL commented on mne/io/artemis123/utils.py:
```diff
@@ -0,0 +1,99 @@
+                locs[vals[0]] = np.array(vals[1::], np.float)
+
+    else:
+        logger.info('Converting Tristan coil file to mne loc file...')
```
Let's not have this in here... we don't want (for various packaging and sanity reasons) to write to the MNE install directory. Let's just distribute it, and leave this generation stuff in here as a comment if you want to keep it, or convert it to a test if you really want to be thorough.

Eric89GXL commented on mne/io/artemis123/utils.py:
```diff
@@ -0,0 +1,99 @@
+    """Load the Coil locations from Tristan CAD drawings."""
+    channel_info = dict()
+    with open(coil_loc_path, 'r') as fid:
+        # skip 2 lines
```
header lines?

Eric89GXL commented on mne/io/artemis123/utils.py:
```diff
@@ -0,0 +1,99 @@
+    R[1, 2] = z_dev[1]
+    R[2, 0] = -z_dev[0]
+    R[2, 1] = -z_dev[1]
+    R[2, 2] = 1 - f * (z_dev[0] * z_dev[0] + z_dev[1] * z_dev[1])
```
Can you put this in `mne.transforms` and add at least some simple tests? For example passing unit vectors should give obvious rotation matrix results.

Eric89GXL replied:
Not worried just thinking ahead this is more generally useful beyond this particular use case. In this sense it fits a bit better in the `mne.transforms` location than here

bloyl replied:
That makes more sense I thought you just wanted me to use an mne.transform as opposed an np.array

Eric89GXL commented:
Oh and one more thing, please add this to `python_reference.rst` and add an entry to `whats_new.rst`

bloyl commented:
Ok, I think I hit most of the code review points. let me know what you think.

Still to do
1) use info['subject_info']
2) add adding some sanity checks to reader. Check numeric values.

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -128,15 +135,20 @@ def _get_artemis123_info(fname):
         meas_date = None
 
+    # build subject info
+    subject_info = {'ID': header_info['Subject ID']}
```
lower case on `'id'`

Eric89GXL commented on mne/io/artemis123/tests/test_artemis123.py:
```diff
@@ -0,0 +1,41 @@
+def test_utils():
+    """Test artemis123 utils."""
+    # make a tempfile
+    fd, path = tempfile.mkstemp()
```
We have a convenience function for this sort of thing. Just do `temp_dir = _TempDir()` and the temporary directory that gets created will automatically be deleted when `temp_dir` gets garbage collected (i.e., at the end of the function call).

Eric89GXL replied:
...and using our version should fix the AppVeyor failure

https://ci.appveyor.com/project/agramfort/mne-python/build/1.0.2404#L822

Eric89GXL commented on mne/transforms.py:
```diff
@@ -247,6 +247,42 @@ def rotation3d(x=0, y=0, z=0):
+        The rotation matrix.
+    """
+    target_z_axis = target_z_axis / np.linalg.norm(target_z_axis)
+    f = 1 / (1 + target_z_axis[2])
```
so you have a problem when `target_z_axis = [0, 0, -1]` right?

Eric89GXL commented:
You have a remaining style error (spelling)

https://travis-ci.org/mne-tools/mne-python/jobs/176058180#L3119

Eric89GXL commented on mne/io/artemis123/tests/test_artemis123.py:
```diff
@@ -28,14 +26,12 @@ def test_data():
+    _generate_mne_locs_file(tmp_fname)
+    installed_locs = _load_mne_locs()
+    generated_locs = _load_mne_locs(tmp_fname)
+    assert_equal(installed_locs, generated_locs)
```
Due to architecture differences on machines, numerical differences could make minor differences here, so something like `assert_allclose(..., atol=1e-7)` would be better

bloyl replied:
unfortunately, installed and generate locs are dicts so assert_allclose doesn't work :(. It seems like only assert_equal supports dicts. Do you have any other ideas?

Eric89GXL replied:
I would just do:
```
assert_equal(set(installed_locs.keys()), set(generated_locs.keys()))
for key in installed_locs.keys():
    assert_allclose(installed_locs[key], generated_locs[keys], atol=1e-7)
```

Eric89GXL commented:
Does this look right to you?

https://3117-1301584-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/auto_examples/visualization/plot_meg_sensors.html

We probably need to add a new helmet.

bloyl commented:
That matches what I am plotting. I still have to compare it the drawings I have from the manufacture, but I think its right.

One thing I remain concerned about is units and how/where they are used through out the code.

For instance the artemis and CTF gradiometers in mne_python have units of T, when I think they should be T/m. I'm also worried that the artemis raw data is in fT so I set the unit_mul where as it seems like your CTF reader is reading in straight T (no unit_mul).

Are there any sanity checks that I can use to convince myself that all the note keeping is correct and being used?

Eric89GXL commented:
> For instance the artemis and CTF gradiometers in mne_python have units of T, when I think they should be T/m.

IIRC some of the sensors get the unit `T` instead of `T/m` because they are synthetic/software gradiometers (difference between two magnetometers separated by a very large distance) that have activity levels more similar to that of magnetometers than gradiometers. So it's mostly a plotting issue... @agramfort does that ring a bell for you, too?

IIRC the only place the units are actually used is for plotting purposes. They aren't used in forward code. Although we do need to be careful in `maxwell_filter` because for gradiometers (synthetic or hardware) the first three external components are necessarily zero...

> I'm also worried that the artemis raw data is in fT so I set the unit_mul where as it seems like your CTF reader is reading in straight T (no unit_mul).

If you do `raw[meg_picks][0]` your data should be on the order of 1e-13-ish. If they are closer to unity, then you should scale them on read or (easier) set `unit_mul`.

agramfort commented:
>
> IIRC some of the sensors get the unit T instead of T/m because they are
> synthetic/software gradiometers (difference between two magnetometers
> separated by a very large distance) that have activity levels more similar
> to that of magnetometers than gradiometers. So it's mostly a plotting
> issue... @agramfort <https://github.com/agramfort> does that ring a bell
> for you, too?
>
> yes it's put in T and not T/m as values are similar to what you get with
neuromag magnetometers in terms of scale. It's because the second
magnetometer in the CTF axial gradiometer is very far from the brain.

bloyl commented:
I just did a quick grep for 'unit_mul' through the mne src tree.

Most of its usage are lines are in io (usually setting it to 0). I don't see any place where it is actually used to scale anything. While in most cases this would just affect the scale of any fit source activity (assuming it maps all the way through to stc objects), you could dream up a case where this multiplier is different for different channels.

So two questions
1) In the artemis reader should I apply the scaling and set the unit_mul = 0? Then artemis data is on par with the other data sources in mne_python
2) Is this something you are worried about in mne_python? Maybe just enforce unit_mul == 0 for all channels?

bloyl commented:
Is there anyone who can take a look at the integration points I added to coil_def.dat?

I based them off of the CTF integration points but it would be good to have someone more in the know look them over.

agramfort commented:
as soon as you read the data raw._data (if preload=True) has been scaled

we have a gain control when writing but no scaling is applied dynamically.
Everything
is scaled on read and stored by default in float64

HTH

Eric89GXL commented:
Hooray the CIs are passing!

> 1) In the artemis reader should I apply the scaling and set the unit_mul = 0? Then artemis data is on par with the other data sources in mne_python

Yeah that sounds right. I thought we used `unit_mul` but actually I was thinking of `scale`, in our code somewhere we do scale by this:
```
info['chs'][k]['cal'] * info['chs'][k]['range']
```
I think the `range` has to do with converting float<->int, so better not to use it. But `cals` is a multiplicative factor you could set to `1e-13` or whatever you need to get the units correct and that would probably work. I'll look into making sure that's not doing something conceptually silly, but it should get you where you need to be without having to do any multiplies yourself manually in the code.

> 2) Is this something you are worried about in mne_python? Maybe just enforce unit_mul == 0 for all channels?

It looks like that is actually the "unit exponent" in the `mne/io/tag.py`, so presumably you could set it to something else if necessary. For now we can just read it and pass it on (as we have been doing) in case some other software sets it meaningfully somewhere.

> Is there anyone who can take a look at the integration points I added to coil_def.dat?

@wronk do you have time to take a quick look at these integration points? You had to deal with integration points most recently I think.

Or @mshamalainen if you have a minute to look at the definitions and changes to `coil_def.dat` that would help, too.

If only we had some phantom data to check the localization quality with... :)

Eric89GXL commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,263 @@
+
+        # other reference channels are unplugged and should be ignored.
+        elif (chan['name'].startswith('REF')):
+            t['coil_type'] = FIFF.FIFFV_COIL_NONE
```
I think we'll want these to still be configured with proper units. For example when concatenating two files, one that had a "bad"/off reference channel, it will be more difficult to do so if we say that one doesn't have a coil type, but the other does. It *should* suffice in the code just to mark it as bad.

@agramfort do you agree? I haven't had to explicitly think about an unplugged channel before, but I guess you can have equivalently flat channels in EEG and those just get marked bad, not set to channel type misc ...

Eric89GXL replied:
Although it does carry some risk to include them in case someone overwrites `info['bads']` without accounting for this.

I think the current approach is fine. In the future we can add an `unplugged='misc' | 'bad'` kwarg to the reader if we really need to support my suggestion (only marking as bad instead of setting channel type to misc).

bloyl replied:
Here is what the manufacture says about those 9 channels.

> The first 2 are bad SQUIDs that were left in the system, but have no coil
attached. The remaining 7 are spares that are shorted, and are useful for
evaluating noise that is not magnetic.

I don't think there is ever any plan of using these channels for anything other then maintenance.

Eric89GXL replied:
Sounds good. Can you add this in a comment?

wronk commented:
> I based them off of the CTF integration points but it would be good to have someone more in the know look them over.

@bloyl, so these are essentially just scaled versions of the CTF integration points?

bloyl commented:
> > I based them off of the CTF integration points but it would be good to have someone more in the know look them over.

> @bloyl, so these are essentially just scaled versions of the CTF integration points?

@wronk, Essentially yes. Using the ctf points i figured out the radius (as a function of the coil radius) and angles of the integration points and then applied that method to the artemis coils. I guess I could have also just scaled the points by the ratio of artemis:ctf coil radii.

Does that seem sensible?

Eric89GXL commented:
@bloyl it might be worth double-checking with Abramowitz and Stegun, see:

https://github.com/mne-tools/mne-python/blob/master/mne/data/coil_def_Elekta.dat#L5

IIRC there is a version you can browse online. Maybe there is some insight there.

wronk commented:
Yes, you should be able to find an old copy online. Unless there is some legacy reason to define coil definitions a certain way, I think it's better to use the corresponding formula from that book for any new coil defs.

Eric89GXL commented:
http://people.math.sfu.ca/~cbm/aands/abramowitz_and_stegun.pdf

Eric89GXL commented:
@bloyl are the Artemis sensors the same as those in the BabySQUID system? If so we don't need new sensor constants 

http://tristantech.com/artemis123babysquid

bloyl commented:
@wronk From looking at  Abramowitz and Stegun, the integration points for CTF and artemis seem to be those suggested. I can double check the numbers but I"m pretty sure they will match.

@Eric89GXL I don't think so. According to the coil_def.dat definition, the babysquid gradiometers are 6mm radius not ~7.5. I also don't see a reference gradiometer, which on artemis is the same coil diameter but a different baseline.

bloyl commented:
@Eric89GXL sorry the baby squid has 6mm diameter coils the artemis has 15mm diameter coils.
http://tristantech.com/wp-content/uploads/2013/09//BabySQUID_description_v6.pdf

Eric89GXL commented:
Is there a similar document for the Artemis123? It would be great to add link comments in `coil_def.dat` with these things

Eric89GXL commented:
BTW @bloyl I cc'ed you an an email with Elekta folks (feel free to correct my baseline error there) but 8001+ sounds like it should work

bloyl commented:
I chose 10001+ because 8001+ and 9001+ are already in the coil_defs.dat as
a 'Sample TMS' and Kriss coils

On Thu, Nov 17, 2016 at 9:39 AM Eric Larson <notifications@github.com>
wrote:

BTW @bloyl <https://github.com/bloyl> I cc'ed you an an email with Elekta
folks (feel free to correct my baseline error there) but 8001+ sounds like
it should work

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/mne-tools/mne-python/pull/3754#issuecomment-261263021>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AFU7nTJVUA59nq4ZAXDKSnGE2R-_sRcVks5q_GcVgaJpZM4Ktxsn>
.

Eric89GXL commented:
Ahh, 10001+ it is, then

wronk commented:
@bloyl, maybe I'm wrong, but the numerical integration doesn't seem to match up. Can you tell me what equations you used from Abramowitz and Stegun for the 10001 sensor (accuracy 2 and 3)?

wronk commented on mne/data/coil_def.dat:
```diff
@@ -501,3 +501,72 @@
+  0.1250    3.033e-03    -5.253e-03     0.000e+00     0.000e+00     0.000e+00     1.000e+00
+  0.1250   -3.033e-03     5.253e-03     0.000e+00     0.000e+00     0.000e+00     1.000e+00
+  0.1250   -3.033e-03    -5.253e-03     0.000e+00     0.000e+00     0.000e+00     1.000e+00
+ -0.2500    0.000e+00     0.000e+00     0.000e+00     0.000e+00     0.000e+00     1.000e+00
```
Should this z-value should be 5.740e-02?

Eric89GXL commented:
@wronk he copied the existing coil def from a CTF system and rescaled it (see comments above)

bloyl commented:
Sorry about the delay I've only gotten to look at the integration points now.

@wronk 
Just looking at the 10001 gradiometer(radius 7.4295E-3 m)
according to eq 25.4.61 of Abramowitz and Stegun

the accuracy 1 (4 points per coil) the points should be +/- h/2, +/- h/2 with weights of 0.25.
so in this case that would be +/- 3.71475E-3, +/- 3.71475E-3 
which is what I have. This also matches the KRISS and baby squid gradiometers.
So I think that one is correct.

Now for accuracy 2 (7 points per coil)
the important numbers are
sqrt(2/3) h = 6.066e-3
sqrt(1/3) h = 3.033e-3
sqrt(2)/2 h = 5.253e-3

which is what is in coil_def.dat file

So I think the 10001 coil should match Abramowitz and Stegun.

What do you think doesn't match?

wronk commented:
Oooh, didn't realize you needed to use the circular integration formula. My brain is stuck on the square formulas because that's what we use with Elekta and the SSS work. Can you put a small comment on top of the `.dat` file specifying A&S equation numbers used for CTF/Artemis/KRISS/Baby Squid coils?

There are a couple individual lines I put comments on, but the overall scheme makes sense to me now.

wronk commented on mne/data/coil_def.dat:
```diff
@@ -501,3 +501,72 @@
+2  10001    2  14  1.4859e-02  5.7404e-02	"Artemis123 axial gradiometer size = 14.8590  mm base = 57.4040  mm"
+  0.2500    0.000e+00     0.000e+00     0.000e+00     0.000e+00     0.000e+00     1.000e+00
+  0.1250    6.066e-03     0.000e+00     0.000e+00     0.000e+00     0.000e+00     1.000e+00
+  0.1250   -6.066e-03     7.429e-19     0.000e+00     0.000e+00     0.000e+00     1.000e+00
```
That looks super useful. Maybe just add a one liner to replace any values in the `int_points` list that are <1e-10 with zero, or something similar.

Eric89GXL commented:
Sounds good to me. A link to the gist in a comment would be useful too

bloyl commented:
@wronk I tried to generalize the file on gist a bit. can you take a look and make sure that things are correct.


@Eric89GXL If you guys think this is genreally useful, is it worth moving into the mne-python or should it just stay a gist?

Eric89GXL commented:
A subsequent PR to formalize it would be great if you have time, but a low priority since (so far) we do this infrequently. But adding refs to the equations in `coil_def.dat` is a good start and could be part of this PR

Eric89GXL commented:
@bloyl from talking to the Elekta folks they want to keep coil constants < 10000, so can you make yours 8500+ instead of 10000+? The idea is that they are similar to BabyMEG so being in the 8000 range makes some sense

Eric89GXL commented:
Or actually 7500+ would be better, since BabyMEG is 7000+

Eric89GXL commented:
Okay you'll need a rebase since I updated `constants.py`. Push your constants to 7501, etc. and I think we should be good to go here, unless you want to add refs to `coil_def.dat` about formulas.

Eric89GXL commented:
I'll push the necessary Artemis123 changes to `constants.py` to `master`. We need them for the Elekta folks.

Eric89GXL commented on mne/tests/test_transforms.py:
```diff
@@ -236,6 +237,25 @@ def test_rotation():
+                [-0.38169517, 0.683832, -0.62183808],
+                [0.75071668, 0.62183808, 0.22302888]]]
+
+    for i, z in enumerate(inp_zs):
```
can just do `for res, z in zip(exp_res, inp_zs)` instead

Eric89GXL commented:
Other than my last few minor comments, LGTM. Almost there!

bloyl commented on mne/io/artemis123/tests/test_artemis123.py:
```diff
@@ -0,0 +1,46 @@
+    raw = read_raw_artemis123(short_no_HPI_fname, preload=True)
+    meg_picks = pick_types(raw.info, meg=True, eeg=False)
+    # checked against matlab reader.
+    assert_almost_equal(raw[meg_picks[12]][0][0][123], 3.0725106596946721e-11)
```
is there a better way to deal with all of that array nesting ?

assert_allclose is fine with me. are the default rtol and atol good?

bloyl commented:
So I think the base IO is done. I have a phantom dataset that I can start to play with once i have some spare time.

the one question i have is.....
1) Do you think the way I handled the unit multipliers makes sense?
basically converting everything to base SI unit and setting the multiplier to 0

Eric89GXL commented:
The `unit_mul` is FIFF_UNITM_NONE (0) for Neuromag so yes that sounds good to me.

I would say let's merge this as-is, and then if we need to make some fixes once you've worked with a phantom then we'll do that in a subsequent PR. Although unlikely, that might end up needing modifications elsewhere (dipole fitting, fwd, etc.).

You probably told me this already, but what kind of phantom data do you have? I assume it's not a 32-dipole Neuromag phantom. If so, we should add an option to `mne.dipole.get_dipole_locations()` for it.

bloyl commented:
the 32-dipole neuromag phantom doesn't fit in the helmet. so we are stuck using a modified CTF saline phantom. the dipole is placed by hand so there isn't much accuracy possible unfortunately. it will let us compare to CTF dipole localization though. I'll probably need some guidance on how to setup simple spherical head models without mris or bems etc.

I'm also going to be starting working on the HPI localization for both artemis and CTF so you aren't free of me just yet.

Eric89GXL commented:
You will probably get good tips from the CTF phantom tutorial:

https://mne-tools.github.io/stable/auto_tutorials/plot_brainstorm_phantom_ctf.html

Eric89GXL commented:
CIs are happy and so am I. +1 for merge (title updated).

@agramfort @jaeilepp feel free to have a look and merge if you're happy

agramfort commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,268 @@
+        if (chan['name'].startswith('MEG')):
+            t['coil_type'] = FIFF.FIFFV_COIL_ARTEMIS123_GRAD
+            t['kind'] = FIFF.FIFFV_MEG_CH
+            # TODO Shouldn't gradiometers be in T/m (Not sure if it matters)
```
can you replace this by a comment explaining why it's set to T?

bloyl replied:
Based on the conversation above, the comment now reads

>             # While gradiometer units are T/m, use T here as values are more
>             # similar to what you get with neuromag magnetometers in terms of
>             # scale due to long baseline.
>             # This choice also matches the one used for CTF

I'm fine with this but can't help but feel somewhat dirty. If we want people to use units and label their plots then we should make sure that the units are correct :)

agramfort commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,268 @@
+        elif (chan['name'] in ref_grad_names):
+            t['coil_type'] = FIFF.FIFFV_COIL_ARTEMIS123_REF_GRAD
+            t['kind'] = FIFF.FIFFV_REF_MEG_CH
+            # TODO Shouldn't gradiometers be in T/m (Not sure if it matters)
```
same thing here

agramfort commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,268 @@
+                             ' channel Types:"%s"' % chan['name'])
+
+        # incorporate unit mulitplier (unit_mul) into calabraion number
+        t['cal'] = t['cal'] * pow(10, (t['unit_mul']))
```
t['cal'] *= 10 ** t['unit_mul']

agramfort commented on mne/io/artemis123/utils.py:
```diff
@@ -0,0 +1,84 @@
+
+def _load_mne_locs(fname=None):
+    """Load MNE locs structure from file (if exists) or recreate it."""
+    if (not fname):
```
if fname is not None:

bloyl replied:
This is a test of whether `fname == None or fname == ''`
basically its a multiline default argument.
I can change it how ever you like
do you prefer
`if not fname:`
or
`if fname is None`

agramfort commented on mne/io/artemis123/utils.py:
```diff
@@ -0,0 +1,84 @@
+            line = line.strip()
+            vals = line.split(',')
+            channel_info[vals[0]] = dict()
+            if (vals[6]):
```
if vals[6]:

agramfort commented on mne/transforms.py:
```diff
@@ -248,6 +248,47 @@ def rotation3d(x=0, y=0, z=0):
 
 
+def rotation3d_align_z_axis(target_z_axis):
+    """Compute a rotaion matrix to align [ 0 0 1] with supplied target z axis.
```
rotaion -> rotation

agramfort commented on mne/transforms.py:
```diff
@@ -248,6 +248,47 @@ def rotation3d(x=0, y=0, z=0):
+
+    Parameters
+    ----------
+    target_z_axis : array, shape = (1, 3)
```
target_z_axis : array, shape (1, 3)

no =

agramfort commented on mne/transforms.py:
```diff
@@ -248,6 +248,47 @@ def rotation3d(x=0, y=0, z=0):
+
+    Returns
+    -------
+    r : array, shape = (3, 3)
```
no =

agramfort commented on mne/transforms.py:
```diff
@@ -248,6 +248,47 @@ def rotation3d(x=0, y=0, z=0):
+    target_z_axis = target_z_axis / np.linalg.norm(target_z_axis)
+    r = np.zeros((3, 3))
+    if ((1 + target_z_axis[2]) < 1E-12):
+        r[0, 0] = 1
```
please 1. not 1 when quantities are meant to be floats

same remake below and above

agramfort commented on mne/transforms.py:
```diff
@@ -248,6 +248,47 @@ def rotation3d(x=0, y=0, z=0):
+
+    # assert that r is a rotation matrix r^t * r = I and det(r) = 1
+    assert(np.any((r.dot(r.T) - np.identity(3)) < 1E-12))
+    assert((np.linalg.det(r) - 1.0) < 1E-12)
```
use det from scipy.linalg

bloyl commented:
Not quite sure how I broke travis. the error seems unrelated to my commit though? is it possibly a configuration thing?

Also I suspect I am wrong about the units thing. My current thought is that the units for 'gradiometers' should be T and not T/m. my thinking is that unless the vendor does a division by the baseline the measured voltage is only sensitive to the difference between coils so the units (after calibration) should be Tesla.

I am confirming this with Tristan engineering but I would rather hold off the MRG until I get some confirmation.

agramfort commented:
I restarted travis

LGTM

units for grads is on paper T/m as it's a \delta B / \delta location (in meters)
I guess you can see the CTF axial gradiometers as denoised magnetometers...

jaeilepp commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,274 @@
+        elif (chan['name'].startswith('MIO')):
+            t['coil_type'] = FIFF.FIFFV_COIL_NONE
+            t['kind'] = FIFF.FIFFV_MISC_CH
+            t['unit'] = FIFF.FIFF_UNIT_V
```
These lines could be condensed to ``if chan['name'].startswith(('AUX', 'TRG', 'MIO')):``.

jaeilepp commented on mne/io/artemis123/artemis123.py:
```diff
@@ -0,0 +1,274 @@
+            raise ValueError('Channel does not match expected' +
+                             ' channel Types:"%s"' % chan['name'])
+
+        # incorporate unit mulitplier (unit_mul) into calabraion number
```
-> ``multiplier (unit_mul) into calibration``

bloyl commented:
Can anyone point me to where in mne/forward the leadfield computation is? I'd like to see how/if the baseline for gradiometers is used.

Eric89GXL commented:
Grep for `"base"` or `'base'`. I don't think it's used in practice, just the integration points given by `rmag`, `cosmag`, and `w`.

We populate a dict based on the coil defs here:

https://github.com/mne-tools/mne-python/blob/master/mne/forward/_make_forward.py#L117

Eventually the heavy lifting is done here:

https://github.com/mne-tools/mne-python/blob/master/mne/forward/_compute_forward.py#L753

bloyl commented:
Ok,

So the measurements of the sensors are in T as the sensors are simply reporting the difference between the two coils with *NO* division by the baseline. Also it doesn't appear that the baselines are incorporated into the forward modeling. So I think we are good.

The fact that no baseline division is used in the forward modeling does suggest that if ever there was real gradiometer data (really units of T/m), the inverse modeling would be incorrect. so hopefully that doesn't happen :)

Sorry for being so pedantic.

agramfort commented:
@Eric89GXL any idea why travis complains?

besides LGTM.

@Eric89GXL feel free to merge if you're happy

Eric89GXL commented:
There is a spelling error that gets caught, I'll fix that in `master`. The other two errors are happening across all PRs, looks like it's due to Anaconda breaking `tvtk` :(

Eric89GXL commented:
Thanks @bloyl!

bloyl commented:
Thanks to you guys for all of the help.

On Tue, Dec 6, 2016 at 10:52 AM Eric Larson <notifications@github.com>
wrote:

Thanks @bloyl <https://github.com/bloyl>!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/mne-tools/mne-python/pull/3754#issuecomment-265186336>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AFU7nQa-wrmY5mIqQts0BSc1W5UqKHiCks5rFYTAgaJpZM4Ktxsn>
.
